AsyncDisplayKit/Details/ASDataController.mm

```diff
@@ -200,25 +216,27 @@ - (void)_layoutNodesFromContexts:(NSArray<ASIndexedNodeContext *> *)contexts ofK
         dispatch_semaphore_signal(sema);
       });
       dispatch_semaphore_wait(sema, DISPATCH_TIME_FOREVER);
-      [self layoutLoadedNodes:subarray fromContexts:[contexts subarrayWithRange:NSMakeRange(j, batchCount)] ofKind:kind];
+      
+      [self _layoutNodes:subarray
+            fromContexts:[contexts subarrayWithRange:NSMakeRange(j, batchCount)]
+                  ofKind:kind];
     } else {
       allocationBlock();
       [_mainSerialQueue performBlockOnMainThread:^{
-        [self layoutLoadedNodes:subarray fromContexts:[contexts subarrayWithRange:NSMakeRange(j, batchCount)] ofKind:kind];
+        [self _layoutNodes:subarray
+              fromContexts:[contexts subarrayWithRange:NSMakeRange(j, batchCount)]
+                    ofKind:kind];
       }];
     }
 
     [allocatedNodes addObjectsFromArray:subarray];
 
     dispatch_group_async(layoutGroup, dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
-      for (NSUInteger k = j; k < j + batchCount; k++) {
-        ASCellNode *node = allocatedNodes[k];
-        // Only measure nodes whose views aren't loaded, since we're in the background.
-        // We should already have measured loaded nodes before we left the main thread, using layoutLoadedNodes:ofKind:atIndexPaths:
-        if (!node.isNodeLoaded) {
-          [self _layoutNode:node withConstrainedSize:contexts[k].constrainedSize];
-        }
-      }
+      // We should already have measured loaded nodes before we left the main thread. Layout the remaining once on a background thread.
+      NSRange range = NSMakeRange(j, batchCount);
+      [self _layoutNodes:[allocatedNodes subarrayWithRange:range]
```

Comment:
I think it would be best to have the method support taking the full arrays, And then pass the range as an additional argument. This pattern is followed in a few other areas, if I recall correctly, and is effective at reducing array copy costs which include memory management calls to retain and release each object in the array.

In the near future, we will start adjusting the size of layout batches and may even provide this as an external API for applications to tune based on the average height of their content, and for smaller batch sizes these copies are likely to add up.

Overall though, this is a really fantastic change, and much welcome to this part of the code base that can benefit greatly from this kind of attention to detail.